Signed pipelines guide docs (#2524)
buildkite/docs: pages/agent/v3/signed_pipelines.md

# Signed pipelines

> > ⚠️ Warning: Signed Pipelines are experimental, and these docs (and the commands and arguments referenced in them) are subject to change. This is also an opt-in feature - if you're keen to get your hands on it, talk to your sales rep.

Reviewer: Suggesting someone reach out to a sales rep is new for the docs and makes this sound very enterprise-y. Do smaller self-service customers have a sales rep? Could we handle this through support instead?

## What are signed pipelines?

> Signed Pipelines is a security feature on the Buildkite Agent that allows you to cryptographically sign steps that are uploaded to Buildkite, and then verify that signature on the Buildkite Agent before running the step. This signature helps agents establish the provenance of the jobs that they are instructed to run - asserting that the steps were uploaded by a trusted agent, and that they haven't been modified since they were uploaded. If an agent detects a signature mismatch, by default, it'll refuse to run the job.

Reviewer: Is it important to mention "by default" here? I'm not sure it's adding much and there's no further mention of changing defaults on the rest of the page.

Author: hmm, we do offer an option to warn instead of block, but you're right, it's never mentioned here. i think removing the "by default" is totally valid

These signatures mean that if a threat actor were able to modify a job in flight, the Agent would refuse to run it due to mismatched signatures.

<details>

>   <summary>I think I've seen this before...</summary>

Reviewer 2: Add an emoji? 🤔

  This work is inspired by the older <a href="https://github.com/buildkite/buildkite-signed-pipeline"><code>buildkite-signed-pipeline</code></a> tool, which was a tool you could add to your agent instances. It had a similar idea - signing steps before they're uploaded to Buildkite, then verifying them when they're run. However, it had a number of limitations, including:
  <ul>
    <li>It had to be installed on every agent instance, leading to more configuration</li>
    <li>It only supported symmetric HS256 signatures, meaning that every verifier could also sign uploads</li>

>     <li>It couldn't sign matrix steps</li>

Reviewer 2: Worth directly linking to matrix doc?

```suggestion
    <li>It couldn't sign [matrix](/docs/pipelines/build-matrix) steps</li>
```

  Many thanks to Seek.com.au, who we collaborated with on the older version of the tool, and whose prior art has been instrumental in the development of this newer version.
</details>

### What we actually sign

As mentioned above, Signed Pipelines establishes that important aspects of steps haven't been changed since pipeline upload. In more detail, we sign the following aspects of steps:

- The command
- Environment variables _as defined in the pipeline YAML_ - environment variables set by the agent, by hooks, or by the user's shell are not signed and can potentially override the environment that a step's command is started with.

- Any plugins used by the step, and their configurations
- The matrix configuration
  + The matrix configuration is signed as a whole, rather than each individual matrix job. This means that the signature will be the same for each job in the matrix. When signatures are verified for matrix jobs, we double-check that the job we've received is a valid construction of the matrix, and that the signature matches the matrix configuration.

> - The step's key, if present [WIP]

Reviewer: What's the go with everything labelled "WIP?" Is it to communicate what work is coming next? If so, what's the timeline you're looking at and how confident are we in it?

(I would generally avoid documenting anything "coming soon," but worth considering in the context if you think it's important)

Additionally, we also sign the following aspects of the build for each job:

- The pipeline's UUID [WIP]
- The organization UUID [WIP]
- The build number [WIP]
- The repo the build is running for [WIP]
- The commit the build is running for [WIP]

## Enabling signed pipelines on your agents

### 1. Generating a key pair
Behind the scenes, Signed Pipelines uses [JSON Web Signing (JWS)](https://datatracker.ietf.org/doc/html/rfc7797) to generate its signatures. This means that you'll need to generate a [JSON Web Key Set (JWKS)](https://datatracker.ietf.org/doc/html/rfc7517) to sign and verify your pipelines with, then configure your agents to use those keys.

Luckily, the Buildkite Agent has you covered! There's a JWKS generation tool built into the agent, which you can use to generate a key pair for you. To run it, you'll need to [install the agent on your machine](/docs/agent/v3/installation), and then run:

```bash
buildkite-agent tool keygen --alg <algorithm> --key-id <key-id>
```

> replacing `<algorithm>` and `<key-id>` with the algorithm you want to use, and the key ID you want to use. For example, to generate an EdDSA key pair with a key ID of `my-key-id`, you'd run:

Reviewer: Is there a link we could include for EdDSA? Maybe just https://en.wikipedia.org/wiki/EdDSA ?

```bash
buildkite-agent tool keygen --alg EdDSA --key-id my-key-id
```

The agent will then generate two JSON Web Key Sets for you, one private and one public, and output them to files in the current directory. You can then use these keys to sign and verify your pipelines, respectively.

If no key id is provided, the agent will generate a random one for you.

> Note that the value of `--alg` must be a valid [JSON Web Signing Algorithm](https://datatracker.ietf.org/doc/html/rfc7518#section-3). The agent does not support the `none` algorithm, or any signature algorithms in the `RS` series (any signing algorithms that use RSASSA-PKCS1 v1.5 signatures, `RS256` for example).

Reviewer: Is there a link we can include for RSASSA-PKCS1 v1.5?

<details>
  <summary>Why doesn't the agent support RSASSA-PKCS1 v1.5 signatures?</summary>

>   In short, it's because RSASSA-PKCS1 v1.5 signatures are generally regarded to be less secure than the newer RSA-PSS signatures. While RSASSA-PKCS1 v1.5 signatures are still largely known to be relatively secure, we want to encourage our users to use the most secure algorithms possible, so when using RSA keys, we only support RSA-PSS signatures. We also recommend looking into ECDSA and EdDSA signatures, which are generally regarded to be more secure than RSA signatures.

Reviewer 2: We could add a reference to RFC 8017 (https://www.rfc-editor.org/rfc/rfc8017.html#section-8) as further justification , particularly, "Although no attacks are known against RSASSA-PKCS1-v1_5, in the interest of increased robustness, RSASSA-PSS is REQUIRED in new applications."

>   In short, it's because RSASSA-PKCS1 v1.5 signatures are generally regarded to be less secure than the newer RSA-PSS signatures. While RSASSA-PKCS1 v1.5 signatures are still largely known to be relatively secure, we want to encourage our users to use the most secure algorithms possible, so when using RSA keys, we only support RSA-PSS signatures. We also recommend looking into ECDSA and EdDSA signatures, which are generally regarded to be more secure than RSA signatures.

Reviewer: This language is hedging our bets a bunch. Could we be more certain, or is it important to keep it loose?

Maybe:

> In short, RSASSA-PKCS1 v1.5 signatures are less secure than the newer RSA-PSS signatures. While RSASSA-PKCS1 v1.5 signatures are still relatively secure, we want to encourage our users to use the most secure algorithms possible, so when using RSA keys, we only support RSA-PSS signatures. We also recommend looking into ECDSA and EdDSA signatures, which are more secure than RSA signatures.

When using asymmetric signing algorithms, we recommend having multiple disjoint pools of agents, each using a different [queue](/docs/agent/v3/queues). One pool should be the _uploaders_ and have access to the private keys. Another pool should be the _runners_ and have access to the public keys. This creates a security boundary between the agents that upload and sign pipelines and the agents that run jobs and verify signatures.

Regarding your specific algorithm choice, any of the supported asymmetric signing algorithms are fine and will be secure. If you're not sure which one to use, `EdDSA` is proven to be secure, has a modern design, wasn't designed by a Nation State Actor, and produces nice short signatures.

For more information on what algorithms the agent supports, run:

```sh
buildkite-agent tool keygen --help
```

### 2. Configure the agents

Next, you need to configure your agents to use the keys you generated. On agents that upload pipelines, add the following to the agent's config file:

```ini

> signing-jwks-file=<path to private keyset>

Reviewer: The linter is picking this up as an error since you're using `keyset` as one word. A quick Google makes me think it's right—you normally see `keySet()` as a method name for a "set of keys." I think we should make it `key set` in each place it's used. Thoughts?

Author: yeah, sweet as - will put in a patch

# Signed pipelines

> 🚧 Experimental feature
> Signed pipelines are experimental, and these docs (and the commands and arguments referenced in them) are subject to change. This is also an opt-in feature, so if you'd like to try using signed pipelines, reach out to [support](https://buildkite.com/support).

> Signed pipelines are a security feature where pipelines are cryptographically signed when uploaded to Buildkite. Agents then verify the signature before running the job. If an agent detects a signature mismatch, it'll refuse to run the job.

Reviewer 3: Secuirty->Overview
![image](https://github.com/buildkite/docs/assets/78014112/852a7bf7-3f39-43ba-bfb4-204adc71a81f)

Might be worth checking this area of the docs also to see if you need to make additions for Signed Pipelines

> Signed pipelines are a security feature where pipelines are cryptographically signed when uploaded to Buildkite. Agents then verify the signature before running the job. If an agent detects a signature mismatch, it'll refuse to run the job.

Reviewer 3: I had these sorts of statements in my old PR. Apologies if you have already covered it: 

> We know that builds and deploys can be run on highly privileged machines, and that an attacker convincing that machine to run a malicious command could compromise production infrastructure.
> As customer secuirty is paramount to Buildkite, and we wish to ensure that all customers are immune by default from our control plane being compromised.

> Signed pipelines allow users to sign the pipelines they upload, which means that agents can refuse to run jobs that are not signed by a trusted public key. This makes it impossible for Buildkite (or an attacker thereof) to tamper with the workload that Buildkite orchestrates onto customer infrastructure.

> Sensitive data, such as source code and secrets, remain within your own environment and are not seen by Buildkite. We are aware that many customers invest heavily in configuring their agents to reject jobs that don't match some set of expectations, such as pipeline/repo/branch filtering, limiting to in-repo scripts not arbitrary commands etc.

> Signed pipelines are a security feature where pipelines are cryptographically signed when uploaded to Buildkite. Agents then verify the signature before running the job. If an agent detects a signature mismatch, it'll refuse to run the job.

Reviewer: Maybe adding something like:

```suggestion
Signed pipelines are a security feature where pipelines are cryptographically signed when uploaded to Buildkite. Agents then verify the signature before running the job. If an agent detects a signature mismatch, it'll refuse to run the job.

Maintaining a strong security boundary is important to Buildkite and informs how we design features. It's also a key reason people choose Buildkite over other CI/CD tools. Signing pipelines improves your security posture by ensuring agents don't run jobs where a malicious actor has modified the instructions. This moves you towards zero-trust CI/CD by further isolating you from Buildkite itself being compromised.
```

  <p>Many thanks to <a href="seek.com.au">SEEK</a>, who we collaborated with on the older version of the tool, and whose prior art has been instrumental in the development of this newer version.</p>
</details>

## Pipeline signatures

Pipeline signatures establish that important aspects of steps haven't been changed since they were uploaded.

The following fields are included in the signature for each step:

- **Commands.**
- **Environment variables defined in the pipeline YAML.** Environment variables set by the agent, hooks, or the user's shell are not signed and can override the environment a step's command is started with.
- **Plugins and plugin configuration.**
- **Matrix configuration.** The matrix configuration is signed as a whole rather than each individual matrix job. This means the signature is the same for each job in the matrix. When signatures are verified for matrix jobs, the agent double-checks that the job it received is a valid construction of the matrix and that the signature matches the matrix configuration.

> - **The step key.** If there is a step key present. [WIP]

Reviewer: Still want to check what WIP means here

Author: in this case it's "we're working on it and we expect to merge it prior to this PR, but don't want to forget to add it to the docs"

- The pipeline's UUID [WIP]
- The organization UUID [WIP]
- The build number [WIP]
- The repo the build is running for [WIP]
- The commit the build is running for [WIP]

## Enabling signed pipelines on your agents

You'll need to configure your agents and update pipeline definitions to enable signed pipelines.

### 1. Generate a key pair

Behind the scenes, signed pipelines use [JSON Web Signing (JWS)](https://datatracker.ietf.org/doc/html/rfc7797) to generate signatures. You'll need to generate a [JSON Web Key Set (JWKS)](https://datatracker.ietf.org/doc/html/rfc7517) to sign and verify your pipelines with, then configure your agents to use those keys.

Luckily, the agent has you covered! A JWKS generation tool is built into the agent, which you can use to generate a key pair for you. To use it, you'll need to [install the agent on your machine](/docs/agent/v3/installation), and then run:

```bash
buildkite-agent tool keygen --alg <algorithm> --key-id <key-id>
```

Replacing the following:

- `<algorithm>` with the signing algorithm you want to use.
- `<key-id>` with the key ID you want to use.

> For example, to generate an EdDSA key pair with a key ID of `my-key-id`, you'd run:

Reviewer: I still think it would be good to add a link on "EdDSA"

Author: https://en.wikipedia.org/wiki/EdDSA should be fine i think, though we go into a bit more detail about EdDSA in particular a bit further down

```bash
buildkite-agent tool keygen --alg EdDSA --key-id my-key-id
```

The agent generates two JWKS in your current directory: one private and one public. You can then use these keys to sign and verify your pipelines.

If no key ID is provided, the agent will generate a random one for you.

Note that the value of `--alg` must be a valid [JSON Web Signing Algorithm](https://datatracker.ietf.org/doc/html/rfc7518#section-3). The agent does not support the `none` algorithm, or any signature algorithms in the `RS` series (any signing algorithms that use RSASSA-PKCS1 v1.5 signatures, `RS256` for example).

<details>
  <summary>Why doesn't the agent support RSASSA-PKCS1 v1.5 signatures?</summary>

>   <p>In short, RSASSA-PKCS1 v1.5 signatures are less secure than the newer RSA-PSS signatures. While RSASSA-PKCS1 v1.5 signatures are still relatively secure, we want to encourage our users to use the most secure algorithms possible, so when using RSA keys, we only support RSA-PSS signatures. We also recommend looking into ECDSA and EdDSA signatures, which are more secure than RSA signatures.</p>

Reviewer: I went with the stronger language here, but happy to go back

Author: this is fine, i think - the reason i hedged so much is because i wanted to avoid cryptography nerds telling me i was wrong, but i'm _pretty sure_ everything written in there is either objectively correct, or a matter of interpretation

</details>

#### Algorithm options

Signed pipelines support many different signing algorithms, which you can simplify into two categories:

- **Symmetric**—where signing and verification use the same keys.

> - **Asymmetric**—where signing and verification use different keys.

Reviewer 3: loooove it

> We recommend using an asymmetric signing algorithm, as it creates a permissions boundary between the agents that upload and sign pipelines, and agents that run other jobs and can verify signatures. This means that if an attacker compromises an agent with verification keys, they can't modify the steps without detection.

Reviewer 3: 🔥🔥🔥🔥🔥🔥

Regarding your specific algorithm choice, any of the supported asymmetric signing algorithms are fine and will be secure. If you're not sure which one to use, `EdDSA` is proven to be secure, has a modern design, wasn't designed by a Nation State Actor, and produces nice short signatures.

For more information on what algorithms the agent supports, run:

```sh
buildkite-agent tool keygen --help
```

### 2. Configure the agents

Next, you need to configure your agents to use the keys you generated. On agents that upload pipelines, add the following to the agent's config file:

```ini
job-signing-jwks-path=<path to private keyset>
job-signing-key-id=<the key id you generated earlier>
job-verification-jwks-path=<path to public keyset>
```

This ensures that whenever those agents upload steps to Buildkite, they'll generate signatures using the private key you generated earlier. It also ensures that those agents verify the signatures of any steps they run, using the public key.

On instances that verify jobs, add:

```ini
job-verification-jwks-path=<path to verification keys>
```

> ### 3. Sign all steps

Reviewer: I tried to avoid introducing new terms with "configured" and "uploaded" steps. Hopefully this works

1. Paste the step definitions into a file on your machine.
1. Generate a new YAML definition with signatures by running:

    ```sh
    buildkite-agent pipeline upload --jwks-file-path <path-to-private-keys> --signing-key-id <key-id> --dry-run --format yaml <path-to-pipeline-file>
    ```

    Replacing the following:
    * `<path-to-private-keys>` with the path to your private keys.
    * `<key-id>` with the key ID.
    * `<path-to-pipeline-file>` with the path to the step definitions you saved from the Buildkite dashboard.

    This outputs the pipeline YAML with the signatures added to the steps.

1. Copy the updated step definitions back into the editor in the Buildkite dashboard.
1. Save the new definition.

For example, if you had a file called `pipeline.yaml` in the current directory, and you wanted to sign it with the key generated earlier, you'd run:

```bash
buildkite-agent pipeline upload --jwks-file-path /path/to/private/key.json --signing-key-id my-key-id --dry-run --format yaml pipeline.yaml
```

## Rotating signing keys

Because signed pipelines use JWKS as their key format, rotating keys is easy.

To rotate your keys:

1. Generate a new key pair.

> 1. Add the new keys to your existing key sets. Be careful not to mix public and private keys.

Reviewer 3: ```suggestion
2. Add the new keys to your existing key sets. Be careful not to mix public and private keys.
```

Reviewer: `1.` is correct for both list items. It displays as the right number when rendered and helps make future refactoring simpler. 😊

![CleanShot 2023-11-21 at 09 59 16](https://github.com/buildkite/docs/assets/119824349/44d7d3a9-cc9a-4010-9a5a-9d4b0a46e179)

Reviewer 3: awesome thank you for the clarity :) I'll admit that I did not have it rendered when making this suggestion!

## Enabling signed pipelines on your agents

You'll need to configure your agents and update pipeline definitions to enable signed pipelines.

### 1. Generate a key pair

Behind the scenes, signed pipelines use [JSON Web Signing (JWS)](https://datatracker.ietf.org/doc/html/rfc7797) to generate signatures. You'll need to generate a [JSON Web Key Set (JWKS)](https://datatracker.ietf.org/doc/html/rfc7517) to sign and verify your pipelines with, then configure your agents to use those keys.

Luckily, the agent has you covered! A JWKS generation tool is built into the agent, which you can use to generate a key pair for you. To use it, you'll need to [install the agent on your machine](/docs/agent/v3/installation), and then run:

```bash
buildkite-agent tool keygen --alg <algorithm> --key-id <key-id>
```

Replacing the following:

- `<algorithm>` with the signing algorithm you want to use.
- `<key-id>` with the key ID you want to use.

For example, to generate an EdDSA key pair with a key ID of `my-key-id`, you'd run:

```bash
buildkite-agent tool keygen --alg EdDSA --key-id my-key-id
```

> The agent generates two JWKS in your current directory: one private and one public. You can then use these keys to sign and verify your pipelines.

Author: the reason i didn't use the acronym here is that JWKS isn't plural (it stands for JSON Web Key Set), and we're generating two JWKSes, and it's a difficult word to pluralize 😅

Author: perfect!

Regarding your specific algorithm choice, any of the supported asymmetric signing algorithms are fine and will be secure. If you're not sure which one to use, `EdDSA` is proven to be secure, has a modern design, wasn't designed by a Nation State Actor, and produces nice short signatures.

For more information on what algorithms the agent supports, run:

```sh
buildkite-agent tool keygen --help
```

### 2. Configure the agents

Next, you need to configure your agents to use the keys you generated. On agents that upload pipelines, add the following to the agent's config file:

```ini
job-signing-jwks-path=<path to private keyset>
job-signing-key-id=<the key id you generated earlier>
job-verification-jwks-path=<path to public keyset>
```

This ensures that whenever those agents upload steps to Buildkite, they'll generate signatures using the private key you generated earlier. It also ensures that those agents verify the signatures of any steps they run, using the public key.

On instances that verify jobs, add:

```ini
job-verification-jwks-path=<path to verification keys>
```

### 3. Sign all steps

> So far, you've configured agents to sign and verify any steps they upload and run. But there's another source of steps: the Buildkite dashboard. Rather than defining steps in your repository, you can also define steps in a pipeline's settings. The most common use for this is to have a single step that uploads a pipeline definition from [a YAML file in the repository](/docs/pipelines/defining-steps#step-defaults-pipeline-dot-yml-file). These steps should also be signed.

Author: > Rather than defining steps in your repository, you can also define steps in a pipeline's settings.

this makes it seem a bit like it's an either or, but in 99.99% of cases, it's going to be both, which is why i introduced the terms... they're both of equal importance when it comes to pipelines, both signed and not. 

maybe replacing this paragraph with something like:

```suggestion
So far, you've configured agents to sign and verify any steps they upload and run. However, in Buildkite, steps are defined both in your repo, and in the Buildkite interface itself. The most common use for the steps in buildkite is a single step that uploads a pipeline definition from [a YAML file in the repository](/docs/pipelines/defining-steps#step-defaults-pipeline-dot-yml-file). The steps stored in buildkite will also need to be signed.
```

.... or something, i think there's still some magic that someone better at writing could put on that para 😅

Reviewer: Good call. How about this version?

```suggestion
So far, you've configured agents to sign and verify any steps they upload and run. However, you also define steps in a pipeline's settings through the Buildkite dashboard. For example, teams commonly use a single step in the Buildkite dashboard to upload a pipeline definition from [a YAML file in the repository](/docs/pipelines/defining-steps#step-defaults-pipeline-dot-yml-file). These steps should also be signed.
```

To sign steps configured in the Buildkite dashboard, you need to add static signatures to the YAML:

1. Copy the existing YAML steps definition from the editor in the Builkdite dashboard.
1. Paste the step definitions into a file on your machine.
1. Generate a new YAML definition with signatures by running:

    ```sh
    buildkite-agent pipeline upload --jwks-file-path <path-to-private-keys> --signing-key-id <key-id> --dry-run --format yaml <path-to-pipeline-file>
    ```

    Replacing the following:
    * `<path-to-private-keys>` with the path to your private keys.
    * `<key-id>` with the key ID.
    * `<path-to-pipeline-file>` with the path to the step definitions you saved from the Buildkite dashboard.

    This outputs the pipeline YAML with the signatures added to the steps.

1. Copy the updated step definitions back into the editor in the Buildkite dashboard.
1. Save the new definition.

For example, if you had a file called `pipeline.yaml` in the current directory, and you wanted to sign it with the key generated earlier, you'd run:

```bash

> buildkite-agent pipeline upload --jwks-file-path /path/to/private/key.json --signing-key-id my-key-id --dry-run --format yaml pipeline.yaml

Author: potentially we should switch it so that they're both the same and _not_ `pipeline.yaml`, eg

```suggestion
For example, if you had a file called `configured-steps.yaml` in the current directory, and you wanted to sign it with the key generated earlier, you'd run:
```bash
buildkite-agent pipeline upload --jwks-file-path /path/to/private/key.json --signing-key-id my-key-id --dry-run --format yaml configured-steps.yaml
```

to avoid confusion with the _actual_ pipeline.yaml, which is getting signed when it gets uploaded 😅 there's a lot of overlapping nomenclature here...

potentially `configured-steps.yaml` no longer makes sense as we're not introducing that term anymore, but i think it should be something that isn't `pipeline.yaml`
